### ISSUE TITLE
kill() / signal() doesn't really work.

### ISSUE THREAD

**Issue by hyc** [msys2/Cygwin#176]
Currently MSYS isn't able to send signals to ordinary Win32 programs. A long time ago I wrote a patch for MSYS1 to accomplish this. https://sourceforge.net/p/mingw/mailman/message/10716023/

It would need refreshing to handle 64bit and 32bit binaries but otherwise should still work. Is there any interest in supporting this?

**Comment by lazka** [comment]
Something like this was implemented in https://github.com/msys2/msys2-runtime/pull/31 if I'm not mistaken.

**Comment by hyc** [comment]
That patch only fixes Ctrl-C handling in a console window. The patch I linked to above handles SIGINT/SIGQUIT/SIGKILL/SIGTERM and works even if the target process isn't attached to a terminal.

**Comment by lazka** [comment]
> That patch only fixes Ctrl-C handling in a console window. The patch I linked to above handles SIGINT/SIGQUIT/SIGKILL/SIGTERM and works even if the target process isn't attached to a terminal.

The patch I linked mentioned them all, and it was written before we used conpty, so also without a attached console.

I don't know much about this, but I think we need an example that you want to work, but currently doesn't, to see if anything is missing.

**Comment by hyc** [comment]
I see, didn't read the entire patch thoroughly before. For the moment the only problem I seem to be having is that SIGTERM just calls ExitProcess() so it bypasses any signal handlers.

**Comment by lazka** [comment]
Just wondering, is there any Windows software using SIGTERM handlers? And if yes, why?

**Comment by sskras** [comment]
@lazka, maybe that [MSDN page](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/signal?view=msvc-170#requirements:~:text=The%20SIGILL%20and%20SIGTERM%20signals%20aren%27t%20generated%20under%20Windows.%20They%27re%20included%20for%20ANSI%20compatibility.%20Therefore%2C%20you%20can%20set%20signal%20handlers%20for%20these%20signals%20by%20using%20signal%2C%20and%20you%20can%20also%20explicitly%20generate%20these%20signals%20by%20calling%20raise.) could shed some light:

> The `SIGILL` and `SIGTERM` signals aren't generated under Windows. They're included for ANSI compatibility. Therefore, you can set signal handlers for these signals by using **`signal`**, and you can also explicitly generate these signals by calling [raise](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/raise?view=msvc-170).

OTOH there might be some F/LOSS that got ported to Windows + CRT and are historically still trying to handle these two.

**Comment by hyc** [comment]
>OTOH there might be some F/LOSS that got ported to Windows + CRT and are historically still trying to handle these two.

Exactly. `SIGTERM` is also the default generated by the `kill` command.

**Comment by lazka** [comment]
> OTOH there might be some F/LOSS that got ported to Windows + CRT and are historically still trying to handle these two.

This seems like a bug to me, or potentially problematic if code is executed then which is not meant for that platform.

**Comment by sskras** [comment]
@lazka, that was a hypothetical case, and I didn't generated it whole.

The part I omitted might be a guarding thread/process that logs the stack traces of the remaining threads/child processes, and then sends SIGTERM (in platform-dependent ways) to them.

Of course, having a working example would make things clearer :)

**Comment by dscho** [comment]
> That patch only fixes Ctrl-C handling in a console window.

That is distinctly untrue. That patch is not even necessary in a Console window. That patch was specifically to get `SIGINT` handlers working in MinTTY, without any Console, not even a pseudo console.

TBH I consider #31 to have addressed this ticket before it was opened 😃

**Comment by hyc** [comment]
> TBH I consider https://github.com/msys2/msys2-runtime/pull/31 to have addressed this ticket before it was opened 😃

Mostly. There are still a few signals that the CRT supports that aren't addressed in #31, so the remaining question is if anyone else cares enough about those missing signals. SIGTERM is most significant to me but others might be useful once in a while.

**Comment by dscho** [comment]
It has been pointed out that the "missing signals" are not produced by Windows. I see very, very little point, then, to produce them in the MSYS2 runtime.

The only reason to handle `SIGTERM` in a special way is to give the `atexit()` handlers a chance to run, which the existing code (as of #31) does.

**Comment by hyc** [comment]
Still broken. Using kill with (HUP/TERM) on a process just shows "Hangup" or "Terminated" on the console, and ignores any signal handlers that the program set. For HUP the program's exit status is 129, for TERM the exit status is 143. If the signal handlers had been invoked, the program would have shut down in an orderly fashion and exited with 0 status.

This prevents OpenLDAP slapd from shutting down cleanly, and triggers alarms in management scripts when it should have just exited cleanly and quietly.
